src/app/clusters/group-key-mgmt-server/group-key-mgmt-server.cpp

```diff
@@ -0,0 +1,410 @@
+
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it;
+    uint16_t fabricIndex = 0;
+    uint16_t keySet      = 0;
+
+    for (it = sGroupKeys.begin(); it != sGroupKeys.end(); ++it)
+    {
+        fabricIndex = std::get<0>(it->first);
+        keySet      = std::get<1>(it->first);
+
+        emberAfGroupKeyManagementClusterPrint("  fabric:%u, set:%u", fabricIndex, keySet);
+        uint8_t index     = it->second->current < GROUP_KEY_SET_SIZE ? it->second->current : 0;
+        RootKey & current = it->second->epoch[index];
+        if (current.active)
+        {
+            emberAfGroupKeyManagementClusterPrint("    current: key(%u), start(%u), policy:%u", sizeof(current.value),
+                                                  sizeof(current.startTime), current.policy);
+
+            index              = previousEpoch(index);
+            RootKey & previous = it->second->epoch[index];
+            if (previous.active)
+            {
+                emberAfGroupKeyManagementClusterPrint("    previous: key(%u), start(%u), policy:%u", sizeof(previous.value),
+                                                      sizeof(previous.startTime), previous.policy);
+
+                index         = previousEpoch(index);
+                RootKey & old = it->second->epoch[index];
+                if (old.active)
+                {
+                    emberAfGroupKeyManagementClusterPrint("    old: key(%u), start(%u), policy:%u", sizeof(old.value),
+                                                          sizeof(old.startTime), old.policy);
+                }
+            }
+        }
+    }
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+static KeySet * getKeySet(uint16_t fabricIndex, uint16_t keySetIndex)
+{
+    std::tuple<uint16_t, uint16_t> index(fabricIndex, keySetIndex);
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it = sGroupKeys.find(index);
+    return it == sGroupKeys.end() ? nullptr : it->second;
+}
+
+static EmberAfStatus setKey(uint16_t fabricIndex, uint16_t keySetIndex, chip::ByteSpan & keyRoot, chip::ByteSpan & epochStartTime,
+                            uint8_t securityPolicy)
+{
+    if (keyRoot.size() != GROUP_KEY_ROOT_SIZE || epochStartTime.size() != GROUP_KEY_EPOCH_SIZE)
+    {
+        return EMBER_ZCL_STATUS_INVALID_VALUE;
+    }
+    else
+    {
+        KeySet * set = getKeySet(fabricIndex, keySetIndex);
+        if (set)
+        {
+            set->current = nextEpoch(set->current);
+        }
+        else
+        {
+            set                                                                  = new KeySet();
+            sGroupKeys[std::tuple<uint16_t, uint16_t>(fabricIndex, keySetIndex)] = set;
+        }
+        RootKey & key = set->epoch[set->current];
+        key.active    = true;
+        memcpy(key.value, keyRoot.data(), GROUP_KEY_ROOT_SIZE);
+        memcpy(key.startTime, epochStartTime.data(), GROUP_KEY_EPOCH_SIZE);
```

Comment:
This assumes a byte order matching between wire format and platform, given that starTime is a scalar. This looks like it is not endian-safe